### PR TITLE
add server side tracing

### PR DESCRIPTION
**Pull request by yangzhouhan**

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -268,6 +276,9 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 			}
 			return
 		}
+		if EnableTracing {
```
Comment (dsymonds) [inline server.go]: Check `traceInfo.tr != nil` instead. Keep only one place here consulting the `EnableTracing` global.

File: server.go [hunk 2]
```diff
@@ -247,7 +248,14 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
 	return t.Write(stream, p, opts)
 }
 
-func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) {
+func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc, srvn string) {
+	var traceInfo traceInfo
+	if EnableTracing {
+		traceInfo.tr = trace.New("Recv."+methodFamily(srvn), srvn)
+		defer traceInfo.tr.Finish()
```
Comment (dsymonds) [inline server.go]: something below should be marking the trace as an error if an error occurs.

File: server.go [hunk 3]
```diff
@@ -247,7 +248,14 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
 	return t.Write(stream, p, opts)
 }
 
-func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) {
+func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc, srvn string) {
```
Comment (dsymonds) [inline server.go]: does this really need an extra arg? is it possible to reconstitute the string from MethodDesc, or some other way?

File: stream.go [hunk 4]
```diff
@@ -282,6 +282,9 @@ type serverStream struct {
 	codec      Codec
 	statusCode codes.Code
 	statusDesc string
+	tracing    bool
```
Comment (dsymonds) [inline stream.go]: format this a bit more like the corresponding `clientStream` fields.

File: server.go (hunk 3 above)
Comment (yangzhouhan) [inline server.go]: Yes, I think an extra arg is necessary here, Method Desc only contains method information, no service name.

**Comment by yangzhouhan** [comment]
fixed, PTAL. Thx.

File: server.go (hunk 3 above)
Comment (iamqizhao) [inline server.go]: stream.Method() should contain that info. You can parse that string to get service name.

File: server.go [hunk 5]
```diff
@@ -303,18 +323,27 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
...
+				traceInfo.tr.LazyLog(&payload{sent: true, msg: reply}, true)
+			}
 		default:
 			panic(fmt.Sprintf("payload format to be supported: %d", pf))
 		}
 	}
 }
 
-func (s *Server) processStreamingRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, sd *StreamDesc) {
+func (s *Server) processStreamingRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, sd *StreamDesc, srvn string) {
```
Comment (iamqizhao) [inline server.go]: ditto

**Comment by yangzhouhan** [comment]
fixed.

File: server.go [hunk 6]
```diff
@@ -248,6 +249,22 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
...
+		traceInfo traceInfo
+		err       error
+	)
+	if EnableTracing {
+		traceInfo.tr = trace.New("Recv."+methodFamily(stream.Method()), stream.Method())
+		defer traceInfo.tr.Finish()
+		traceInfo.firstLine.client = false
+		traceInfo.tr.LazyLog(&traceInfo.firstLine, false)
+		defer func() {
+			if err != nil {
```
Comment (iamqizhao) [inline server.go]: if err != nil && err != io.EOF ?

File: server.go [hunk 7]
```diff
@@ -248,6 +249,22 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
...
+				traceInfo.tr.LazyLog(&fmtStringer{"%v", []interface{}{err}}, true)
+				traceInfo.tr.SetError()
+			}
+		}()
+	}
 	p := &parser{s: stream}
 	for {
 		pf, req, err := p.recvMsg()
```
Comment (iamqizhao) [inline server.go]: this actually creates a new "err".

File: server.go [hunk 8]
```diff
@@ -292,40 +310,68 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 			}
 			if err := s.sendResponse(t, stream, reply, compressionNone, opts); err != nil {
 				if _, ok := err.(transport.ConnectionError); ok {
-					return
+					return err
 				}
 				if e, ok := err.(transport.StreamError); ok {
 					statusCode = e.Code
 					statusDesc = e.Desc
+					return err
```
Comment (iamqizhao) [inline server.go]: fuse this with the following return.

File: server.go [hunk 9]
```diff
@@ -292,40 +310,68 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
...
 				} else {
 					statusCode = codes.Unknown
 					statusDesc = err.Error()
+					return err
 				}
 			}
-			t.WriteStatus(stream, statusCode, statusDesc)
+			return t.WriteStatus(stream, statusCode, statusDesc)
+			if traceInfo.tr != nil {
```
Comment (iamqizhao) [inline server.go]: move this upwards.

File: server.go (hunk 7 above)
Comment (dsymonds) [inline server.go]: That's fine, as long as the return statements below are not naked (as this commit also changes). The defer above will still work.

File: server.go [hunk 10]
```diff
@@ -247,13 +248,26 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
...
-func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) {
+func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) (err error) {
+	var traceInfo traceInfo
+	if EnableTracing {
+		traceInfo.tr = trace.New("Recv."+methodFamily(stream.Method()), stream.Method())
+		defer traceInfo.tr.Finish()
+		traceInfo.firstLine.client = false
+		traceInfo.tr.LazyLog(&traceInfo.firstLine, false)
+		defer func() {
+			if err != nil || err != io.EOF {
```
Comment (dsymonds) [inline server.go]: &&

File: server.go [hunk 11]
```diff
@@ -292,40 +310,68 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
...
+				ss.traceInfo.tr.SetError()
+				ss.mu.Unlock()
+			}
+		}()
 	}
 	if appErr := sd.Handler(srv.server, ss); appErr != nil {
 		if err, ok := appErr.(rpcError); ok {
 			ss.statusCode = err.code
 			ss.statusDesc = err.desc
+			return err
```
Comment (iamqizhao) [inline server.go]: do not return this error which is not an op error.

File: server.go [hunk 12]
```diff
@@ -292,40 +310,68 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
...
 		} else {
 			ss.statusCode = convertCode(appErr)
 			ss.statusDesc = appErr.Error()
+			return nil
```
Comment (iamqizhao) [inline server.go]: have a single return for this and the above branch.

File: stream.go [hunk 13]
```diff
@@ -309,6 +330,20 @@ func (ss *serverStream) SendMsg(m interface{}) error {
...
 
-func (ss *serverStream) RecvMsg(m interface{}) error {
+func (ss *serverStream) RecvMsg(m interface{}) (err error) {
+	defer func() {
+		if ss.tracing {
+			ss.mu.Lock()
+			if ss.traceInfo.tr != nil {
+				if err == nil {
+					ss.traceInfo.tr.LazyLog(&payload{sent: false, msg: m}, true)
+				} else {
```
Comment (dsymonds) [inline stream.go]: is this going to catch io.EOF?

File: server.go (hunk 7 above)
Comment (iamqizhao) [inline server.go]: @dsymonds 

FYI, my comment was before we decided to add a return value (error) for this function. github is bad on showing this ...

**Comment by iamqizhao** [comment]
LGTM. 

I will merge after getting @dsymonds LGTM and green travis.

File: server.go [hunk 14]
```diff
@@ -247,13 +248,27 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
...
-func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) {
+func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) (err error) {
+	var traceInfo traceInfo
+	if EnableTracing {
+		traceInfo.tr = trace.New("Recv."+methodFamily(stream.Method()), stream.Method())
+		defer traceInfo.tr.Finish()
+		traceInfo.firstLine.client = false
+		traceInfo.tr.LazyLog(&traceInfo.firstLine, false)
+		defer func() {
+			// The trace only log the first operation err and dosen't log the application error
```
Comment (dsymonds) [inline server.go]: drop this comment. It has typos, and the code below is self-evident.

File: server.go [hunk 15]
```diff
@@ -283,38 +301,63 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
...
+		s:       stream,
+		p:       &parser{s: stream},
+		codec:   s.opts.codec,
+		tracing: EnableTracing,
+	}
+	if ss.tracing {
+		ss.traceInfo.tr = trace.New("Recv."+methodFamily(stream.Method()), stream.Method())
+		ss.traceInfo.firstLine.client = false
+		ss.traceInfo.tr.LazyLog(&ss.traceInfo.firstLine, false)
+		defer func() {
```
Comment (dsymonds) [inline server.go]: merge this with the next defer

File: server.go [hunk 16]
```diff
@@ -351,6 +396,7 @@ func (s *Server) handleStream(t transport.ServerTransport, stream *transport.Str
 	}
 	// Unary RPC or Streaming RPC?
 	if md, ok := srv.md[method]; ok {
+
```
Comment (dsymonds) [inline server.go]: remove this

File: stream.go [hunk 17]
```diff
@@ -300,7 +307,21 @@ func (ss *serverStream) SetTrailer(md metadata.MD) {
 	return
 }
 
-func (ss *serverStream) SendMsg(m interface{}) error {
+func (ss *serverStream) SendMsg(m interface{}) (err error) {
+	defer func() {
+		if ss.tracing {
+			ss.mu.Lock()
+			if ss.traceInfo.tr != nil {
```
Comment (dsymonds) [inline stream.go]: if `ss.tracing` is true, this should always be true.

Or is there a way that SendMsg can get called after the stream is finished?

File: stream.go [hunk 18]
```diff
@@ -309,6 +330,22 @@ func (ss *serverStream) SendMsg(m interface{}) error {
...
-func (ss *serverStream) RecvMsg(m interface{}) error {
+func (ss *serverStream) RecvMsg(m interface{}) (err error) {
+	defer func() {
+		if ss.tracing {
+			ss.mu.Lock()
+			if ss.traceInfo.tr != nil {
+				if err == nil {
+					ss.traceInfo.tr.LazyLog(&payload{sent: false, msg: m}, true)
+				} else {
+					if err != io.EOF {
```
Comment (dsymonds) [inline stream.go]: turn the previous line into an `else if` with this.

**Comment by yangzhouhan** [comment]
fixed.

File: server.go [hunk 19]
```diff
@@ -283,38 +300,59 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
...
+		codec:   s.opts.codec,
+		tracing: EnableTracing,
+	}
+	if ss.tracing {
+		ss.traceInfo.tr = trace.New("Recv."+methodFamily(stream.Method()), stream.Method())
+		ss.traceInfo.firstLine.client = false
+		ss.traceInfo.tr.LazyLog(&ss.traceInfo.firstLine, false)
+		defer func() {
+			ss.mu.Lock()
+			if err != nil {
```
Comment (dsymonds) [inline server.go]: if err != nil && err != io.EOF {

**Comment by dsymonds** [comment]
LGTM after addressing that one conditional.

**Comment by iamqizhao** [comment]
seems break end2end_test. reverting ... investigate tomorrow.

**Pull request merged**